### ISSUE TITLE
Default values not working

### ISSUE THREAD

**Issue by rmortes** [lumeland/cms#22]
I'm trying out Lume. I'm liking it so far, but I found a small problem. Default field values aren't working

<details>
<summary>Version</summary>
"lume/": "https://deno.land/x/lume@v2.2.3/"
"lume/cms/": "https://cdn.jsdelivr.net/gh/lumeland/cms@0.5.5/"
</details>

None of the following configs make a default value appear on the frontend

```ts
{
  name: "date",
  type: "datetime",
  label: "Published date",
  value: new Date(),
  description: "Set a future date if you want to publish it later",
  attributes: {
    placeholder: "For example: 2024-01-01 00:00:01",
  },
},

{
  name: "content",
  type: "markdown",
  label: "Page content",
  value: "Write **markdown** code here",
  description:
    `<a target="_blank" href="https://www.markdownguide.org">More info about markdown syntax</a>`,
},


{
  name: "createdAt",
  type: "datetime",
  mode: "create",
},
{
  name: "updatedAt",
  type: "datetime",
  mode: "update",
},
```

If it changes anything, I'm trying to use DenoKV to store my data, but I've also tried with the disk driver and it's the same behavior. Am I doing something wrong here? Thanks!

**Comment by rmortes** [comment]
If it helps, here's the outerHTML code generated by each of those examples

<details>
<summary>Example 1</summary>

```html
<f-datetime data-nameprefix="changes" data-field="{&quot;tag&quot;:&quot;f-datetime&quot;,&quot;label&quot;:&quot;Published date&quot;,&quot;cmsContent&quot;:{&quot;basePath&quot;:&quot;/admin&quot;,&quot;site&quot;:{&quot;name&quot;:&quot;Lume CMS&quot;,&quot;url&quot;:&quot;http://localhost:3000/&quot;},&quot;collections&quot;:{&quot;Entradas del Blog&quot;:{&quot;name&quot;:&quot;Entradas del Blog&quot;,&quot;permissions&quot;:{&quot;create&quot;:true,&quot;delete&quot;:true}}},&quot;documents&quot;:{},&quot;uploads&quot;:{}},&quot;name&quot;:&quot;date&quot;,&quot;type&quot;:&quot;datetime&quot;,&quot;value&quot;:&quot;2024-09-11T10:14:17.705Z&quot;,&quot;description&quot;:&quot;Set a future date if you want to publish it later&quot;,&quot;attributes&quot;:{&quot;placeholder&quot;:&quot;For example: 2024-01-01 00:00:01&quot;}}" class="field">
    <label for="field_changes.date">Published date</label><div class="field-description">Set a future date if you want to publish it later</div><input placeholder="For example: 2024-01-01 00:00:01" id="field_changes.date" name="changes.date" type="datetime-local" class="input" style="--max-width: none;"></f-datetime>
```

</details>

<details>
<summary>Example 2</summary>

```html
<f-markdown data-nameprefix="changes" data-field="{&quot;tag&quot;:&quot;f-markdown&quot;,&quot;label&quot;:&quot;Page content&quot;,&quot;cmsContent&quot;:{&quot;basePath&quot;:&quot;/admin&quot;,&quot;site&quot;:{&quot;name&quot;:&quot;Lume CMS&quot;,&quot;url&quot;:&quot;http://localhost:3000/&quot;},&quot;collections&quot;:{&quot;Entradas del Blog&quot;:{&quot;name&quot;:&quot;Entradas del Blog&quot;,&quot;permissions&quot;:{&quot;create&quot;:true,&quot;delete&quot;:true}}},&quot;documents&quot;:{},&quot;uploads&quot;:{}},&quot;name&quot;:&quot;contento&quot;,&quot;type&quot;:&quot;markdown&quot;,&quot;value&quot;:&quot;Write **markdown** code here&quot;,&quot;description&quot;:&quot;<a target=\&quot;_blank\&quot; href=\&quot;https://www.markdownguide.org\&quot;>More info about markdown syntax</a>&quot;}" class="field">
    <label for="field_changes.0">Page content</label><div class="field-description"><a target="_blank" href="https://www.markdownguide.org">More info about markdown syntax</a></div><textarea id="field_changes.contento" name="changes.contento" hidden=""></textarea><div class="tools"><div class="tools-group"><button class="buttonIcon" type="button"><u-icon name="text-b"><svg xmlns="http://www.w3.org/2000/svg" width="32" height="32" fill="#000000" viewBox="0 0 256 256"><path d="M170.48,115.7A44,44,0,0,0,140,40H72a8,8,0,0,0-8,8V200a8,8,0,0,0,8,8h80a48,48,0,0,0,18.48-92.3ZM80,56h60a28,28,0,0,1,0,56H80Zm72,136H80V128h72a32,32,0,0,1,0,64Z"></path></svg></u-icon></button><button class="buttonIcon" type="button"><u-icon name="text-italic"><svg xmlns="http://www.w3.org/2000/svg" width="32" height="32" fill="#000000" viewBox="0 0 256 256"><path d="M200,56a8,8,0,0,1-8,8H157.77L115.1,192H144a8,8,0,0,1,0,16H64a8,8,0,0,1,0-16H98.23L140.9,64H112a8,8,0,0,1,0-16h80A8,8,0,0,1,200,56Z"></path></svg></u-icon></button><button class="buttonIcon" type="button"><u-icon name="text-strikethrough"><svg xmlns="http://www.w3.org/2000/svg" width="32" height="32" fill="#000000" viewBox="0 0 256 256"><path d="M224,128a8,8,0,0,1-8,8H175.93c9.19,7.11,16.07,17.2,16.07,32,0,13.34-7,25.7-19.75,34.79C160.33,211.31,144.61,216,128,216s-32.33-4.69-44.25-13.21C71,193.7,64,181.34,64,168a8,8,0,0,1,16,0c0,17.35,22,32,48,32s48-14.65,48-32c0-14.85-10.54-23.58-38.77-32H40a8,8,0,0,1,0-16H216A8,8,0,0,1,224,128ZM76.33,104a8,8,0,0,0,7.61-10.49A17.3,17.3,0,0,1,83.11,88c0-18.24,19.3-32,44.89-32,18.84,0,34.16,7.42,41,19.85a8,8,0,0,0,14-7.7C173.33,50.52,152.77,40,128,40,93.29,40,67.11,60.63,67.11,88a33.73,33.73,0,0,0,1.62,10.49A8,8,0,0,0,76.33,104Z"></path></svg></u-icon></button></div><div class="tools-group"><button class="buttonIcon" type="button"><u-icon name="text-h-one"><svg xmlns="http://www.w3.org/2000/svg" width="32" height="32" fill="#000000" viewBox="0 0 256 256"><path d="M152,56V176a8,8,0,0,1-16,0V124H48v52a8,8,0,0,1-16,0V56a8,8,0,0,1,16,0v52h88V56a8,8,0,0,1,16,0Zm75.77,49a8,8,0,0,0-8.21.39l-24,16a8,8,0,1,0,8.88,13.32L216,127V208a8,8,0,0,0,16,0V112A8,8,0,0,0,227.77,105Z"></path></svg></u-icon></button><button class="buttonIcon" type="button"><u-icon name="text-h-two"><svg xmlns="http://www.w3.org/2000/svg" width="32" height="32" fill="#000000" viewBox="0 0 256 256"><path d="M152,56V176a8,8,0,0,1-16,0V124H48v52a8,8,0,0,1-16,0V56a8,8,0,0,1,16,0v52h88V56a8,8,0,0,1,16,0Zm88,144H208l33.55-44.74a32,32,0,1,0-55.73-29.93,8,8,0,1,0,15.08,5.34,16.28,16.28,0,0,1,2.32-4.3,16,16,0,1,1,25.54,19.27L185.6,203.2A8,8,0,0,0,192,216h48a8,8,0,0,0,0-16Z"></path></svg></u-icon></button><button class="buttonIcon" type="button"><u-icon name="text-h-three"><svg xmlns="http://www.w3.org/2000/svg" width="32" height="32" fill="#000000" viewBox="0 0 256 256"><path d="M152,56V176a8,8,0,0,1-16,0V124H48v52a8,8,0,0,1-16,0V56a8,8,0,0,1,16,0v52h88V56a8,8,0,0,1,16,0Zm73.52,90.63,21-30A8,8,0,0,0,240,104H192a8,8,0,0,0,0,16h32.63l-19.18,27.41A8,8,0,0,0,212,160a20,20,0,1,1-14.29,34,8,8,0,1,0-11.42,11.19A36,36,0,0,0,248,180,36.07,36.07,0,0,0,225.52,146.63Z"></path></svg></u-icon></button><button class="buttonIcon" type="button"><u-icon name="text-h-four"><svg xmlns="http://www.w3.org/2000/svg" width="32" height="32" fill="#000000" viewBox="0 0 256 256"><path d="M152,56V176a8,8,0,0,1-16,0V124H48v52a8,8,0,0,1-16,0V56a8,8,0,0,1,16,0v52h88V56a8,8,0,0,1,16,0Zm96,112h-8V144a8,8,0,0,0-16,0v24H195.1l20.49-61.47a8,8,0,0,0-15.18-5.06l-24,72A8,8,0,0,0,184,184h40v24a8,8,0,0,0,16,0V184h8a8,8,0,0,0,0-16Z"></path></svg></u-icon></button></div><div class="tools-group"><button class="buttonIcon" type="button"><u-icon name="link-simple"><svg xmlns="http://www.w3.org/2000/svg" width="32" height="32" fill="#000000" viewBox="0 0 256 256"><path d="M165.66,90.34a8,8,0,0,1,0,11.32l-64,64a8,8,0,0,1-11.32-11.32l64-64A8,8,0,0,1,165.66,90.34ZM215.6,40.4a56,56,0,0,0-79.2,0L106.34,70.45a8,8,0,0,0,11.32,11.32l30.06-30a40,40,0,0,1,56.57,56.56l-30.07,30.06a8,8,0,0,0,11.31,11.32L215.6,119.6a56,56,0,0,0,0-79.2ZM138.34,174.22l-30.06,30.06a40,40,0,1,1-56.56-56.57l30.05-30.05a8,8,0,0,0-11.32-11.32L40.4,136.4a56,56,0,0,0,79.2,79.2l30.06-30.07a8,8,0,0,0-11.32-11.31Z"></path></svg></u-icon></button><a class="buttonIcon" href="https://www.markdownguide.org/basic-syntax/" target="_blank"><u-icon name="question"><svg xmlns="http://www.w3.org/2000/svg" width="32" height="32" fill="#000000" viewBox="0 0 256 256"><path d="M140,180a12,12,0,1,1-12-12A12,12,0,0,1,140,180ZM128,72c-22.06,0-40,16.15-40,36v4a8,8,0,0,0,16,0v-4c0-11,10.77-20,24-20s24,9,24,20-10.77,20-24,20a8,8,0,0,0-8,8v8a8,8,0,0,0,16,0v-.72c18.24-3.35,32-17.9,32-35.28C168,88.15,150.06,72,128,72Zm104,56A104,104,0,1,1,128,24,104.11,104.11,0,0,1,232,128Zm-16,0a88,88,0,1,0-88,88A88.1,88.1,0,0,0,216,128Z"></path></svg></u-icon></a></div></div></f-markdown>
```

</details>

<details>
<summary>Example 3</summary>

```html
<f-datetime data-nameprefix="changes" data-field="{&quot;tag&quot;:&quot;f-datetime&quot;,&quot;label&quot;:&quot;Created At&quot;,&quot;cmsContent&quot;:{&quot;basePath&quot;:&quot;/admin&quot;,&quot;site&quot;:{&quot;name&quot;:&quot;Lume CMS&quot;,&quot;url&quot;:&quot;http://localhost:3000/&quot;},&quot;collections&quot;:{&quot;Entradas del Blog&quot;:{&quot;name&quot;:&quot;Entradas del Blog&quot;,&quot;permissions&quot;:{&quot;create&quot;:true,&quot;delete&quot;:true}}},&quot;documents&quot;:{},&quot;uploads&quot;:{}},&quot;name&quot;:&quot;createdAt&quot;,&quot;type&quot;:&quot;datetime&quot;,&quot;mode&quot;:&quot;create&quot;}" class="field">
    <label for="field_changes.createdAt">Created At</label><input id="field_changes.createdAt" name="changes.createdAt" type="datetime-local" class="input" style="--max-width: none;"></f-datetime>
<f-datetime data-nameprefix="changes" data-field="{&quot;tag&quot;:&quot;f-datetime&quot;,&quot;label&quot;:&quot;Updated At&quot;,&quot;cmsContent&quot;:{&quot;basePath&quot;:&quot;/admin&quot;,&quot;site&quot;:{&quot;name&quot;:&quot;Lume CMS&quot;,&quot;url&quot;:&quot;http://localhost:3000/&quot;},&quot;collections&quot;:{&quot;Entradas del Blog&quot;:{&quot;name&quot;:&quot;Entradas del Blog&quot;,&quot;permissions&quot;:{&quot;create&quot;:true,&quot;delete&quot;:true}}},&quot;documents&quot;:{},&quot;uploads&quot;:{}},&quot;name&quot;:&quot;updatedAt&quot;,&quot;type&quot;:&quot;datetime&quot;,&quot;mode&quot;:&quot;update&quot;}" class="field">
    <label for="field_changes.updatedAt">Updated At</label><input id="field_changes.updatedAt" name="changes.updatedAt" type="datetime-local" class="input" style="--max-width: none;"></f-datetime>
```

</details>

**Comment by oscarotero** [comment]
Okay, let me see.
Default values is something that is not fully resolved in the CMS.

**Comment by oscarotero** [comment]
I've released the version 0.6.0:
- The `mode` option for `date` and `datetime` fields was removed. It wasn't a good design decision.
- There's the new `current_datetime` field that you can use to update automatically the document with the current date when it's saved.
- The `value` option now works fine and it's used as the default value for new documents.

**Comment by rmortes** [comment]
DAMN that was fast O.o
I'll check it out ASAP and let you know if it works correctly

**Comment by rmortes** [comment]
Already tested. It's almost working perfectly now. I can set the createdAt value correctly, and the updatedAt is stored correctly as well, but not displays the correct information. Let me elaborate. I'm using the following setup:

´´´ts
{
  name: "createdAt",
  type: "datetime",
  attributes: {
    "readonly": true,
  },
  init(field) {
    field.value = new Date();
  },
},
{
  name: "updatedAt",
  type: "current_datetime",
  attributes: {
    "readonly": true,
  },
},
´´´

I can create a new document, and edit it as well, and the database will always contain the correct createdAt value and the correct updatedAt value. However, the UI will display the updatedAt value as the current datetime, regardless of the database value. I understand why this happens, I think. To keep the updatedAt value up to date, you just change the value in the form to whatever the current datetime is, so that when you save the form, the value gets updated as well.

However, if a less tech-savvy client sees this, they'll be pretty confused about why this is happening.

I think the solution would be to display the database value in the form, but to send the current datetime whenever the form is updated. 

This can get rid of another edge case I just noticed. Since you set the updatedAt value to the current datetime when the page is loaded, if I take my time editing the collection and save it, the updatedAt value will not be correct, it'll be however minutes it took me to make the edit earlier than it should.

I don't know if currently your implementation allows for this. If you don't have the time required to make this changes, maybe I can chime in and lend a hand? I'll probably end up contributing to the repo either way since I hope to start using this CMS in production pretty soon

**Comment by oscarotero** [comment]
You're right, the `current_datetime` field generates the value when the form is loaded, because the following reasons:
- If the previous value is shown, users may think this value is not updated automatically.
- Sometimes, the user want set a different value because reasons.
- It would be confused to send the form with a value but a different value is stored.

But you have some valid points too. So maybe the behavior can be improved. Some ideas:
- The input shows the current time (as it does now).
- But every minute (or second), the value is updated (like a clock)
- If the value is changed by the user, it's not longer updated automatically.
- Maybe, we can show somewhere the previous value, just in case the user want to keep the previous value.

What do you think?

If you like the idea and want to work on that, it's just editing [the web component file](https://github.com/lumeland/cms/blob/main/static/components/f-current_datetime.js).

**Comment by rmortes** [comment]
I've opened a pull request for this issue (https://github.com/lumeland/cms/pull/23)

I've changed the current_datetime component to behave the way you suggested:

- The previous value is shown below the input (the wording is terrible, I'm a backend dev what can I say?)
- The value is updated automatically (each second) to reflect the current datetime
- The value stops updating once the field gets "dirty" (the user updates it manually)

Additionally, I've added a new component, readonly_current_datetime, that behaves how I personally think it should:

- The field is readonly to let the user know they are not supposed to change it in any way
- The value corresponds to the last time the collection was modified
- The value is updated just before the form is sent, so it changes locally for the user (they can see the collection was updated "just now") and the new value is sent with the form so the database value changes as well

Honestly, the name "readonly_current_datetime" seems terrible to me. I'm thinking "updated_datedate" or "updated_at_datetime" would be better, since this is a pretty opinionated field and the name reflects that. Wdyt?
